wasmcloud-chat/actors/messages/src/lib.rs

```diff
@@ -1,9 +1,172 @@
+extern crate serde;
+
 extern crate wapc_guest as guest;
-use wasmcloud_actor_core as actor;
+extern crate wasmcloud_actor_core as actor;
+extern crate wasmcloud_actor_eventstreams as streams;
+extern crate wasmcloud_actor_messaging as broker;
+extern crate wasmcloudchat_messages_interface as messages;
 
+use chrono::prelude::*;
 use guest::prelude::*;
+use messages::{ChannelMessage, ProcessAck};
+use std::collections::HashMap;
+use url::Url;
+
+mod cloudevent;
+
+use cloudevent::CloudEvent;
+
+const WCCHAT_SCHEME: &str = "wcchat";
+const EVENT_TYPE_MESSAGE_PUBLISHED: &str = "com.wasmcloud.chat.events.messagepublished";
+const EVENT_SOURCE: &str = "/actors/messaging";
+
+const HMK_ORIGIN_CHANNEL: &str = "origin_channel";
+const HMK_ORIGIN_USER: &str = "origin_user";
+const HMK_TIMESTAMP: &str = "timestamp";
+const HMK_MESSAGE_TEXT: &str = "messagetext";
+const HMK_ID: &str = "message_id";
+
+const MSGTYPE_MESSAGE: &str = "MSG";
+const VALID_MESSAGE_TYPES: [&str; 1] = [MSGTYPE_MESSAGE];
 
 #[actor::init]
 fn init() {
-    // Register your message handlers here
-}
\ No newline at end of file
+    messages::Handlers::register_process_message(process_message);
+}
+
+// Processing a message:
+// 1 - validate the message
+// 2 - publish it to the appropriate message broker subject
+// 3 - emit event to event stream
+// 4 - ACK
+fn process_message(message: ChannelMessage) -> HandlerResult<ProcessAck> {
+    let target = MessageTarget::from(&message);
+    if let MessageTarget::Unknown(message) = target {
+        return Err(format!("Unable to select target for message: {}", message).into());
+    }
+
+    Ok(validate_message(&message)
+        .and_then(|_| publish_broker_message(&target, &message))
+        .and_then(|_| emit_stream_event(&target, &message))
+        .map_or_else(
+            |_| ProcessAck::success(&message.message_id),
+            |_| {
+                ProcessAck::fail(
+                    &message.message_id,
+                    "Did not publish and emit to event stream",
+                )
+            },
+        ))
```

Comment:
This has got to be one of the cleanest pieces of Rust business logic that exists